alerts/geomodel/locality.py

```diff
@@ -0,0 +1,255 @@
+    '''Produced by calls to functions operating on lists of `State`s to
+    indicate when an update was applied without having to maintain distinct
+    lists.
+    '''
+
+    state: State
+    did_update: bool
+
+    def flat_map(fn: Callable[[State], 'Update'], u: 'Update') -> 'Update':
+        '''Apply a function to a `State` that produces an `Update` against the
+        state contained within an established `Update`.  The resulting `Update`
+        will have its `did_update` field set to `True` if either the original
+        or the new `Update` are `True`.
+        '''
+
+        new = fn(u.state)
+
+        return Update(new.state, u.did_update or new.did_update)
+
+
+JournalInterface = Callable[[Entry, str], None]
+QueryInterface = Callable[[SearchQuery, str], Optional[Entry]]
+
+
+def _dict_take(dictionary, keys):
+    return {key: dictionary[key] for key in keys}
+
+
+def wrap_journal(client: ESClient) -> JournalInterface:
+    '''Wrap an `ElasticsearchClient` in a closure of type `JournalInterface`.
+    '''
+
+    def wrapper(entry: Entry, esindex: str):
+        document = dict(entry.state._asdict())
+
+        client.save_object(
+            index=esindex,
+            body=document,
+            doc_id=entry.identifier)
+
+    return wrapper
+
+
+def wrap_query(client: ESClient) -> QueryInterface:
+    '''Wrap an `ElasticsearchClient` in a closure of type `QueryInterface`.
+    '''
+
+    def wrapper(query: SearchQuery, esindex: str) -> Optional[Entry]:
+        results = query.execute(client, indices=[esindex]).get('hits', [])
+
+        if len(results) == 0:
+            return None
+
+        state_dict = results[0].get('_source', {})
+        try:
+            state_dict['localities'] = [
+                # Convert dictionary localities into `Locality`s after
+                # parsing the `datetime` from `lastaction`.
+                Locality(**_dict_take({
+                    k: v if k != 'lastaction' else _parse_datetime(v)
+                    for k, v in loc.items()
+                }, Locality._fields))
+                for loc in state_dict['localities']
+            ]
+
+            eid = results[0]['_id']
+            state = State(**_dict_take(state_dict, State._fields))
+
+            return Entry(eid, state)
+        except TypeError:
+            return None
+        except KeyError:
+            return None
+
+    return wrapper
+
+
+def from_event(
+        event: Dict[str, Any],
+        radius=_DEFAULT_RADIUS_KM
+) -> Optional[Locality]:
+    '''Extract locality information from an event if it is present in order
+    to produce a `Locality` for which an authenticated action was taken.
+    '''
+
+    _source = event.get('_source', {})
+
+    source_ip = _source.get('sourceipaddress')
+    geo_data = _source.get('sourceipgeolocation')
+
+    if source_ip is None or geo_data is None:
+        return None
+
+    now = datetime.strftime(datetime.utcnow(), '%Y-%m-%dT%H:%M:%S.%f+00:00')
```

Comment:
I think we should use the `toUTC` function https://github.com/mozilla/MozDef/blob/master/mozdef_util/mozdef_util/utilities/toUTC.py#L10  This line then becomes `now = toUTC(datetime.now()).isoformat()`